tests/unit/json_test.c

```diff
@@ -847,6 +847,78 @@ static void test_iterator_current(void)
     JsonDestroy(json);
 }
 
+static bool VisitPrimitive(JsonElement *primitive, void *data)
+{
+    assert_int_equal(JsonGetElementType(primitive), JSON_ELEMENT_TYPE_PRIMITIVE);
+
+    Writer *writer = data;
+    WriterWrite(writer, JsonPrimitiveGetAsString(primitive));
+    WriterWriteChar(writer, ',');
+    return true;
+}
+
+static bool VisitArray(JsonElement *array, void *data)
+{
+    assert_int_equal(JsonGetElementType(array), JSON_ELEMENT_TYPE_CONTAINER);
+    assert_int_equal(JsonGetContainerType(array), JSON_CONTAINER_TYPE_ARRAY);
+
+    Writer *writer = data;
+    WriterWriteF(writer, "[%zd]", JsonLength(array));
+    return true;
+}
+
+static bool VisitObject(JsonElement *object, void *data)
+{
+    assert_int_equal(JsonGetElementType(object), JSON_ELEMENT_TYPE_CONTAINER);
+    assert_int_equal(JsonGetContainerType(object), JSON_CONTAINER_TYPE_OBJECT);
+
+    Writer *writer = data;
+    WriterWriteChar(writer, '{');
+    JsonIterator iter = JsonIteratorInit(object);
+    while (JsonIteratorHasMore(&iter))
+    {
+        WriterWrite(writer, JsonIteratorNextKey(&iter));
+        WriterWriteChar(writer, ',');
+    }
+    WriterWriteChar(writer, '}');
+    return true;
+}
+
+static bool VisitArrayAbortOnEmpty(JsonElement *array, void *data)
+{
+    assert_int_equal(JsonGetElementType(array), JSON_ELEMENT_TYPE_CONTAINER);
+    assert_int_equal(JsonGetContainerType(array), JSON_CONTAINER_TYPE_ARRAY);
+
+    Writer *writer = data;
+    WriterWriteF(writer, "[%zd]", JsonLength(array));
```

Comment:
```suggestion
    WriterWriteF(writer, "[%zd],", JsonLength(array));
```

Comma here would make the trace more readable, IMO.